clippy_lints/src/needless_borrow.rs

```diff
@@ -86,50 +108,69 @@ impl<'tcx> LateLintPass<'tcx> for NeedlessBorrow {
         }
     }
     fn check_pat(&mut self, cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>) {
-        if pat.span.from_expansion() || self.derived_item.is_some() {
-            return;
-        }
         if_chain! {
-            if let PatKind::Binding(BindingAnnotation::Ref, .., name, _) = pat.kind;
-            if let ty::Ref(_, tam, mutbl) = *cx.typeck_results().pat_ty(pat).kind();
-            if mutbl == Mutability::Not;
-            if let ty::Ref(_, _, mutbl) = *tam.kind();
+            if let PatKind::Binding(BindingAnnotation::Ref, id, name, _) = pat.kind;
+            if !in_external_macro(cx.sess(), pat.span);
+            if let ty::Ref(_, tam, _) = *cx.typeck_results().pat_ty(pat).kind();
             // only lint immutable refs, because borrowed `&mut T` cannot be moved out
-            if mutbl == Mutability::Not;
+            if let ty::Ref(_, _, Mutability::Not) = *tam.kind();
             then {
+                let (span, search_loc) = match get_pat_usable_location(cx.tcx, pat) {
+                    Some(x) => x,
+                    None => return,
+                };
+                let ctxt = span.ctxt();
+                if pat.span.ctxt() != ctxt {
+                    // The context of the pattern is different than the context using the binding.
+                    // Changing the pattern might affect other code which needs the ref binding.
+                    return;
+                }
+
+                let mut app = Applicability::MachineApplicable;
+                let mut can_fix = true;
+                let mut lint = NEEDLESS_BORROW;
+                let mut changes = vec![
+                    (pat.span, snippet_with_context(cx, name.span, ctxt, "..", &mut app).0.into_owned()),
+                ];
+
+                foreach_local_usage(cx, id, search_loc, |e| {
```

Comment:
This adds a HIR traversal for every ref binding. Limiting the traversal to the binding scope will reduce it, but only by a constant factor (visiting half of the body on average). So we are visiting the HIR an extra `ref_bindings * 0.5` times here.

I suggest that you have one visitor that visits everything once in `check_body`. The visitor would...
1. Record ref bindings to a hash map
2. Upon encountering a binding usage, remove from the hash map if not lint-able.

Reply:
It only runs when the ref binding is a reference to a reference, which will always trigger the lint.

Reply:
What about a reference to `Copy`? What is the extent of the `can_fix == false` case?

Reply:
I still think we should avoid adding N traversals to the HIR if at all possible.

Reply:
`can_fix` is to deal with an edge case with macros. Should probably add a note about that.

Reply:
References to copy types aren't linted, and I'd rather do that in a different PR.

Reply:
Using `check_body` also doubles the number of traversals. It's likely that will be much larger than the number of ref bindings to reference types.

Reply:
> Using `check_body` also doubles the number of traversals.

It adds exactly one traversal. This adds an unbounded number of traversals since it can traverse the same HIR N times for each binding.

> It's likely that will be much larger than the number of ref bindings to reference types.

Maybe, on average, but we should think in terms of the worst case as well as the average case.

Reply:
Just to add numbers here, this doesn't even occur in any of the crates checked by lintcheck.

Reply:
Hmm that is surprising. Those crates may be skewed against ref bindings, but that is still quite compelling. Alright, I'm convinced.

Reply:
Does rustc ever skip calling the lint pass when all it's lints are allowed? There's actually no reason to use a visitor if it's guaranteed to always call `check_expr`.

Reply:
> Does rustc ever skip calling the lint pass when all it's lints are allowed?

I'm not entirely sure but I think that if the lint warns anywhere for a crate, then all the callbacks will be called. But basically you should be able to rely on the callbacks. So yeah if you can get it to work with no visitor at all, go for it. I guess you could just use callbacks and lint in `post_check_body`? Actually, any reference to `check_body` should probably be `check_item` because of closures.